include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,1244 @@
+              update_preconditioner = update_preconditioner_predicate();
+
+            if (update_preconditioner) // update preconditioner
+              flag = setup_preconditioner(x);
+          }
+
+        return flag;
+      },
+      [&](const VectorType &x, VectorType &v) -> int {
+        Assert(
+          apply_jacobian,
+          ExcMessage(
+            "No apply_jacobian function has been attached to the NOXSolver object."));
+
+        n_jacobian_applications++;
+
+        // apply Jacobian
+        return apply_jacobian(x, v);
+      },
+      [&](const VectorType &f, VectorType &x, const double tolerance) -> int {
+        n_nonlinear_iterations++;
+
+        // invert Jacobian
+        if (solve_with_jacobian)
+          {
+            // without tracking of linear iterations
+            return solve_with_jacobian(f, x, tolerance);
+          }
+        else if (solve_with_jacobian_and_track_n_linear_iterations)
+          {
+            // with tracking of linear iterations
+            const int n_linear_iterations =
+              solve_with_jacobian_and_track_n_linear_iterations(f,
+                                                                x,
+                                                                tolerance);
+
+            if (n_linear_iterations == -1)
+              return 1;
+
+            this->n_last_linear_iterations = n_linear_iterations;
+
+            return 0;
+          }
+        else
+          {
+            Assert(
+              false,
+              ExcMessage(
+                "Neither a solve_with_jacobian or a "
+                "solve_with_jacobian_and_track_n_linear_iterations function "
+                "has been attached to the NOXSolver object."));
+
+            Assert(false, ExcNotImplemented());
+            return 1;
+          }
+      }));
+
+    // setup solver control
+    auto check =
+      Teuchos::rcp(new NOX::StatusTest::Combo(NOX::StatusTest::Combo::OR));
+
+    if (additional_data.abs_tol > 0.0)
+      {
+        const auto additional_data_norm_f_abs =
+          Teuchos::rcp(new NOX::StatusTest::NormF(additional_data.abs_tol));
+        check->addStatusTest(additional_data_norm_f_abs);
+      }
+
+    if (additional_data.rel_tol > 0.0)
+      {
+        const auto additional_data_norm_f_rel = Teuchos::rcp(
+          new NOX::StatusTest::RelativeNormF(additional_data.rel_tol));
+        check->addStatusTest(additional_data_norm_f_rel);
+      }
+
+    if (additional_data.max_iter > 0)
+      {
+        const auto additional_data_max_iterations =
+          Teuchos::rcp(new NOX::StatusTest::MaxIters(additional_data.max_iter));
+        check->addStatusTest(additional_data_max_iterations);
+      }
+
+    if (this->check_iteration_status)
+      {
+        const auto info = Teuchos::rcp(
+          new internal::NOXWrappers::NOXCheck(this->check_iteration_status));
+        check->addStatusTest(info);
+      }
```

Comment:
Since this user-defined `check_iteration_status` is added last to `check` which is based on `OR` logic, it won't be called if either of the previous checks succeeds. Maybe it is better to add it first?